bundles/org.openhab.binding.qbus/src/main/java/org/openhab/binding/qbus/internal/protocol/QbusCommunication.java

```diff
@@ -0,0 +1,913 @@
+     * to all messages coming from Qbus.
+     *
+     * @param addr : IP-address of Qbus Server
+     * @param port : Communication port of Qbus server
+     * @throws InterruptedException
+     * @throws IOException
+     *
+     */
+
+    public synchronized void startCommunication() throws IOException, InterruptedException {
+        QbusBridgeHandler handler = bridgeCallBack;
+        CTDConnected = false;
+
+        if (qEventsRunning) {
+            throw new IOException("Previous listening thread is still active.");
+        }
+
+        if (handler == null) {
+            throw new IOException("No Bridge handler initialised.");
+        }
+
+        InetAddress addr = InetAddress.getByName(handler.getAddress());
+        Integer port = handler.getPort();
+
+        if (port != null) {
+            Socket socket = new Socket(addr, port);
+            qSocket = socket;
+            qOut = new PrintWriter(socket.getOutputStream(), true);
+            qIn = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+            logger.debug("Connected via local port {} from thread {}", socket.getLocalPort(),
+                    Thread.currentThread().getId());
+        } else {
+            return;
+        }
+
+        setSN();
+        getSN();
+
+        // Connect to Qbus server
+        Connect();
+
+        // If Qbus Client is connected then initialize and start listening for incoming commands, else put Bridge
+        // offline
+        if (CTDConnected == true) {
+            initialize();
+            Thread thread = new Thread(this::qEvents);
+            thread.setName("OH-binding");
+            thread.setDaemon(true);
+            thread.start();
+        } else {
+            handler.bridgeOffline("No communication with Qbus client");
+        }
+    }
+
+    /**
+     * Cleanup socket when the communication with Qbus Server is closed.
+     *
+     * @throws IOException
+     *
+     */
+    public synchronized void stopCommunication() {
+        listenerStopped = true;
+
+        Socket socket = qSocket;
+
+        if (socket != null) {
+            try {
+                socket.close();
+            } catch (IOException ignore) {
+                // ignore IO Error when trying to close the socket if the intention is to close it anyway
+            }
+        }
+
+        qSocket = null;
+
+        CTDConnected = false;
+        logger.debug("Communication stopped from thread {}", Thread.currentThread().getId());
+    }
+
+    /**
+     * Close and restart communication with Qbus Server.
+     *
+     * @throws InterruptedException
+     */
+    public synchronized void restartCommunication() {
+        stopCommunication();
+
+        logger.debug("Qbus: restart communication from thread {}", Thread.currentThread().getId());
+
+        try {
+            startCommunication();
+        } catch (InterruptedException | IOException e) {
```

Comment:
The `InterruptedException` is thrown when OH is shut-down. That means the current operation should be finished as fast as possible and no further operations should be done. You can achieve this by not catching the exception here but throw the InterruptedException and catch it at the highest level you can. As this is not an error, it shouldn't be logged, but only handled.

Reply:
You need to remove the `InterruptedException` from the catch clause to let it be thrown.

Reply:
Can you work on this?